### PR TITLE
VPN-4131 - Implement addon session state

### PR DESCRIPTION
**Pull request by brizental**
**This is a large PR, so I suggest going commit by commit during the review.**

One note about [DO NOT MERGE THIS ONE: Refactor split tunneling tutorial to use state](https://github.com/mozilla-mobile/mozilla-vpn-client/commit/afd6e095194e41a47762e146782ea20de5a5ecc2). Turns out we were too late to the party as this addon is actually disabled starting from 2.14... I added a commit with the refactor for illustration purposes, but we can't merge this change because this tutorial will not be enabled in versions that will have addon state support.

- [x] Oh I just realized there are leaks on the AddonSessionState. Gotta fix that before merging as well!

### CODE REVIEW

File: addons/tutorial_04_split_tunneling/vpnIsOn.js
Comment (bakulf) [inline addons/tutorial_04_split_tunneling/vpnIsOn.js:9]: Let's make this file back-compatible. Maybe with `if ("state" in addon) ...`

File: addons/tutorial_04_split_tunneling/vpnWasOn.js
Comment (bakulf) [inline addons/tutorial_04_split_tunneling/vpnWasOn.js:7]: Same here.

File: docs/add-on.md
Comment (bakulf) [inline docs/add-on.md:68]: Local?

File: src/apps/vpn/addons/addon.cpp
Comment (bakulf) [inline src/apps/vpn/addons/addon.cpp:429]: Eventually all of this will be replaced with global states.

File: src/apps/vpn/addons/state/addonstate.h
Comment (bakulf) [inline src/apps/vpn/addons/state/addonstate.h:27]: Why a manual delete? Can you set the parent obj in m_session?

File: src/apps/vpn/addons/state/addonstate.h
Comment (bakulf) [inline src/apps/vpn/addons/state/addonstate.h:49]: Maybe we don't need to allocate it. Just to avoid a nullptr exposed to the code by mistake.

File: src/apps/vpn/addons/state/addonstatebase.cpp
Comment (bakulf) [inline src/apps/vpn/addons/state/addonstatebase.cpp:36]: Here you are creating a hash table, then you populate it, and then you return a copy of it as return value. What about if you receive a reference as param to avoid the extra copy.

File: src/apps/vpn/addons/state/addonstatebase.cpp
Comment (bakulf) [inline src/apps/vpn/addons/state/addonstatebase.cpp:74]: const &

File: src/apps/vpn/addons/state/addonstatebase.cpp
Comment (bakulf) [inline src/apps/vpn/addons/state/addonstatebase.cpp:110]: Return

File: src/apps/vpn/addons/state/addonstatebase.h
Comment (bakulf) [inline src/apps/vpn/addons/state/addonstatebase.h:32]: A CTOR with a parent?

File: src/apps/vpn/addons/state/addonstatebase.h
Comment (bakulf) [inline src/apps/vpn/addons/state/addonstatebase.h:75]: Should this be private?

File: src/apps/vpn/addons/state/addonstatebase.cpp
Comment (bakulf) [inline src/apps/vpn/addons/state/addonstatebase.cpp:7]: Not needed

File: src/apps/vpn/addons/state/addonsessionstate.h
Comment (bakulf) [inline src/apps/vpn/addons/state/addonsessionstate.h:11]: None of these headers are needed.

File: src/apps/vpn/addons/state/addonsessionstate.h
Comment (bakulf) [inline src/apps/vpn/addons/state/addonsessionstate.h:22]: A parent please :)

File: src/apps/vpn/addons/state/addonsessionstate.h
Comment (bakulf) [inline src/apps/vpn/addons/state/addonsessionstate.h:47]: Private.

File: src/apps/vpn/addons/state/addonsessionstate.cpp
Comment (bakulf) [inline src/apps/vpn/addons/state/addonsessionstate.cpp:10]: Remove these headers

File: src/apps/vpn/addons/state/addonsessionstate.cpp
Comment (bakulf) [inline src/apps/vpn/addons/state/addonsessionstate.cpp:12]: Remove this too

File: src/apps/vpn/addons/state/addonsessionstate.cpp
Comment (bakulf) [inline src/apps/vpn/addons/state/addonsessionstate.cpp:21]: Const & + parent

File: src/apps/vpn/addons/state/addonsessionstate.cpp
Comment (bakulf) [inline src/apps/vpn/addons/state/addonsessionstate.cpp:29]: I wonder if m_state should be in the base class. You will need it also for the other implementations. The other impls will have a different propagation algorithm but m_state will probably remain.

File: tests/functional/testAddons.js
Comment (bakulf) [inline tests/functional/testAddons.js:198]: Unrelated?

File: tests/unit/testaddonsessionstate.cpp
Comment (bakulf) [inline tests/unit/testaddonsessionstate.cpp:16]: Without a parent you leak this object.

**Review by bakulf — CHANGES_REQUESTED** [review]
Wow! Well done! Just small comments about the parenting.

File: addons/tutorial_04_split_tunneling/vpnIsOn.js
Comment (brizental) [reply addons/tutorial_04_split_tunneling/vpnIsOn.js:9]: We shouldn't merge this actually... I left this for illustration purposes, I am going to drop this commit before merging. This tutorial is disabled starting from v2.14, so I think there is no point.

File: docs/add-on.md
Comment (brizental) [reply docs/add-on.md:68]: Oops, nice catch!

File: src/apps/vpn/addons/state/addonstate.h
Comment (brizental) [reply src/apps/vpn/addons/state/addonstate.h:27]: Well, because of ignorance really :D Didn't realize that was the reason for setting the parent. Now that you say it, it makes sense :flushed:

File: src/apps/vpn/addons/state/addonstatebase.h
Comment (brizental) [reply src/apps/vpn/addons/state/addonstatebase.h:75]: Just for testing purposes, no. I use it in the "SpyState" class I created for tests.

File: src/apps/vpn/addons/state/addonsessionstate.h
Comment (brizental) [reply src/apps/vpn/addons/state/addonsessionstate.h:11]: How do you know this? I always wonder haha

File: src/apps/vpn/addons/state/addonsessionstate.cpp
Comment (brizental) [reply src/apps/vpn/addons/state/addonsessionstate.cpp:29]: I had it in the bsae class initially. But I am not too sure it will be required for the other states. Like for local state, will we load from settings to memory on initialize? Or will we always just quuery the settings on `set`/`get`/`clear`? Ifwe do that then `m_state` is not necesssary. We can always move it back there if we realize all of them use it.

File: tests/functional/testAddons.js
Comment (brizental) [reply tests/functional/testAddons.js:198]: Oh... Yes, unlreated. Oops.

File: tests/unit/testaddonsessionstate.cpp
Comment (brizental) [reply tests/unit/testaddonsessionstate.cpp:16]: Yes that is why I had the explicit `delete`... But I'll change that.

File: addons/tutorial_04_split_tunneling/vpnIsOn.js
Comment (brizental) [reply addons/tutorial_04_split_tunneling/vpnIsOn.js:9]: Already discussed this offline with @bakulf and he agrees with the above comment :)

File: src/apps/vpn/addons/addon.cpp
Comment (bakulf) [inline src/apps/vpn/addons/addon.cpp:360]: I don't like this too much. Because you are checking something here and then you use it elsewhere extracting the same keys, without checking them.

File: src/apps/vpn/addons/addon.cpp
Comment (bakulf) [inline src/apps/vpn/addons/addon.cpp:422]: For instance you don't check this object at all... This is fine, but inconsistent

File: src/apps/vpn/addons/addon.cpp
Comment (bakulf) [reply src/apps/vpn/addons/addon.cpp:360]: What about if you create the addon state after this block? See the condition block. This way you can use addon as the parent.

**Review by bakulf — CHANGES_REQUESTED** [review]
Nothing wrong. Let's talk about style. ;)

File: src/apps/vpn/addons/addon.cpp
Comment (brizental) [reply src/apps/vpn/addons/addon.cpp:360]: Yes, I agree. I'll change it.

File: src/apps/vpn/addons/addon.h
Comment (bakulf) [inline src/apps/vpn/addons/addon.h:61]: Q_ASSERT(m_state);

File: src/apps/vpn/addons/state/addonstatebase.h
Comment (bakulf) [inline src/apps/vpn/addons/state/addonstatebase.h:30]: = default;

**Review by bakulf — APPROVED** [review]

**Pull request merged**
